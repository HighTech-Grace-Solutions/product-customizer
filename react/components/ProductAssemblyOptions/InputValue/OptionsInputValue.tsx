import React, { FC, useMemo } from 'react'
import { Dropdown } from 'vtex.styleguide'
import { useCssHandles } from 'vtex.css-handles'
import { useIntl } from 'react-intl'

import { useProductAssemblyGroupState } from '../../ProductAssemblyContext/Group'
import useInputValue from './useInputValue'
import OptionBox from './OptionBox'
import {
  formatSubscriptionLabel,
  formatSubscriptionOptions,
  isSubscription,
} from '../../../modules/subscriptions'

const DROPDOWN_OPTIONS_HANDLES = ['optionsInputValueDropdown'] as const
const BOX_OPTIONS_HANDLES = [
  'optionsInputValue',
  'optionsInputValueLabelContainer',
  'optionsInputValueLabel',
  'optionsInputValueOptionBoxContainer',
] as const

export type OptionDisplay = 'select' | 'box' | 'smart'

type Options = {
  label: string
  value: string
}

type InnerProps = {
  label: string
  options: Options[]
  inputId: string
}

type Props = {
  optionsDisplay?: OptionDisplay
  inputValueInfo: OptionsInputValue
}

const DropdownOptions: FC<InnerProps> = ({ label, options, inputId }) => {
  const [state, onChange] = useInputValue(inputId)
  const handles = useCssHandles(DROPDOWN_OPTIONS_HANDLES)

  const handleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    const { value } = e.target

    onChange({ value })
  }

  return (
    <div className={`${handles.optionsInputValueDropdown} mb4`}>
      <Dropdown
        value={state}
        onChange={handleChange}
        label={label}
        options={options}
      />
    </div>
  )
}

const BoxOptions: FC<InnerProps> = ({ label, options, inputId }) => {
  const [state, onChange] = useInputValue(inputId)
  const handles = useCssHandles(BOX_OPTIONS_HANDLES)

  const handleKeyDown = (event: React.KeyboardEvent) => {
    const selected = state as string
    const selectedIndex: number = options.findIndex((i) => i.value === selected)

    switch (event.key) {
      case 'ArrowRight': {
        const count = options.length
        const nextSelectedIndex = (selectedIndex + 1) % count
        const newOption = options[nextSelectedIndex]

        onChange({ value: newOption.value })
        break
      }

      case 'ArrowLeft': {
        const count = options.length
        // Linter is triggering a false positive here :/
        // eslint-disable-next-line @typescript-eslint/restrict-plus-operands
        const previousSelectedIndex = (selectedIndex - 1 + count) % count
        const newOption = options[previousSelectedIndex]

        onChange({ value: newOption.value })
        break
      }

      case 'Home': {
        const [newOption] = options

        onChange({ value: newOption.value })
        break
      }

      case 'End': {
        const count = options.length
        const newOption = options[count - 1]

        onChange({ value: newOption.value })
        break
      }

      default: {
        break
      }
    }
  }

  return (
    <div className={`${handles.optionsInputValue} mb4`}>
      <div className={`${handles.optionsInputValueLabelContainer} mb3`}>
        <span
          className={`${handles.optionsInputValueLabel} c-muted-1 t-small overflow-hidden`}
        >
          {label}
        </span>
      </div>
      <div
        className={`${handles.optionsInputValueOptionBoxContainer} inline-flex flex-wrap flex items-center`}
      >
        {options.map((option) => {
          return (
            <OptionBox
              key={option.value}
              onKeyDown={handleKeyDown}
              option={option.label}
              selected={state === option.value}
              onClick={() => onChange({ value: option.value })}
            />
          )
        })}
      </div>
    </div>
  )
}

const OptionsInputValue: FC<Props> = ({
  optionsDisplay = 'select',
  inputValueInfo,
}) => {
  const intl = useIntl()
  const state = useProductAssemblyGroupState()
  const inputId = inputValueInfo.label
  const inputDomain = inputValueInfo.domain
  const isSubscriptionOption = isSubscription(inputId)

  // We get the current `frequency` value here instead of inside the useMemo
  // so it doesn't depend on the state, as it is a new object at every update.
  // For non-subscription inputValues, this is `undefined` which is fine.
  // Prevents re-rendering when changing `purchaseDay`.
  const frequency = state?.valuesOfInputValues[
    'vtex.subscription.key.frequency'
  ] as string

  const { label, options } = useMemo(() => {
    // not a subscription, just map the options
    if (!isSubscriptionOption) {
      return {
        label: inputId,
        options: inputDomain.map((value) => ({ label: value, value })),
      }
    }

    return {
      label: formatSubscriptionLabel(inputId, intl),
      options: formatSubscriptionOptions({
        inputId,
        inputDomain,
        frequency,
        intl,
      }),
    }
  }, [isSubscriptionOption, inputId, inputDomain, frequency, intl])

  const OptionComponent =
    optionsDisplay === 'box' ? BoxOptions : DropdownOptions

  return (
    <OptionComponent
      inputId={inputValueInfo.label}
      label={label}
      options={options}
    />
  )
}

export default OptionsInputValue
