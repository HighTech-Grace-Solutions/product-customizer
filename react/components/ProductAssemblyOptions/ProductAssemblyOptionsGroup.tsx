import React, { FC, Fragment } from 'react'
import { Button } from 'vtex.styleguide'
import { IOMessage } from 'vtex.native-types'
import { useIntl } from 'react-intl'

import {
  useProductAssemblyGroupState,
  useProductAssemblyGroupDispatch,
} from '../ProductAssemblyContext/Group'
import useAssemblyOptionsModifications from '../../modules/useAssemblyOptionsModifications'
import { ProductAssemblyItemProvider } from '../ProductAssemblyContext/Item'
import ProductAssemblyOptionsItem from './ProductAssemblyOptionsItem'
import {
  formatSubscriptionLabel,
  isSubscription,
} from '../../modules/subscriptions'

interface Props {
  initiallyOpened?: 'always' | 'required'
}

const ProductAssemblyOptionsGroup: FC<Props> = ({
  children,
  initiallyOpened = 'required',
}) => {
  const intl = useIntl()
  const assemblyOptionGroup = useProductAssemblyGroupState() as AssemblyOptionGroupState
  const dispatch = useProductAssemblyGroupDispatch()

  useAssemblyOptionsModifications(assemblyOptionGroup)

  const changeOptinInput = () => {
    dispatch({
      type: 'OPTIN',
      args: {
        groupPath: assemblyOptionGroup.path,
      },
    })
  }

  const groupName = isSubscription(assemblyOptionGroup)
    ? formatSubscriptionLabel(assemblyOptionGroup, intl)
    : assemblyOptionGroup.groupName

  return (
    <Fragment>
      {assemblyOptionGroup.optin === false && initiallyOpened === 'required' ? (
        <Button variation="secondary" onClick={changeOptinInput}>
          <IOMessage
            id="store/product-customizer.add-assembly"
            values={{ name: groupName }}
          />
        </Button>
      ) : (
        <Fragment>
          <div className="flex justify-between">
            <div className="ttc-s pv4 c-muted-2 t-small">{groupName}</div>
            {assemblyOptionGroup.required === false && (
              <div>
                <Button
                  size="small"
                  collapseRight
                  variation="tertiary"
                  onClick={changeOptinInput}
                >
                  <IOMessage
                    id="store/product-customizer.remove-assembly"
                    values={{ name: groupName }}
                  />
                </Button>
              </div>
            )}
          </div>
          {assemblyOptionGroup.items ? (
            Object.values(assemblyOptionGroup.items).map((item) => {
              return (
                <ProductAssemblyItemProvider item={item} key={item.id}>
                  <ProductAssemblyOptionsItem>
                    {children}
                  </ProductAssemblyOptionsItem>
                </ProductAssemblyItemProvider>
              )
            })
          ) : (
            <ProductAssemblyItemProvider item={null}>
              {children}
            </ProductAssemblyItemProvider>
          )}
        </Fragment>
      )}
    </Fragment>
  )
}

export default ProductAssemblyOptionsGroup
