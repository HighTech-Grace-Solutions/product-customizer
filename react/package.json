{
  "private": true,
  "scripts": {
    "lint": "eslint --ext ts,tsx,js,jsx .",
    "lint:fix": "eslint --fix --ext ts,tsx,js,jsx .",
    "test": "vtex-test-tools test"
  },
  "dependencies": {
    "classnames": "^2.2.6",
    "graphql": "^14.6.0",
    "ramda": "^0.26.1",
    "react": "^16.9.0",
    "react-apollo": "^3.1.3",
    "react-dom": "^16.9.0",
    "react-intl": "^4.2.2",
    "slugify": "^1.3.5",
    "smoothscroll-polyfill": "^0.4.3"
  },
  "devDependencies": {
    "@apollo/react-testing": "^3.1.3",
    "@testing-library/react-hooks": "^2.0.1",
    "@types/classnames": "^2.2.9",
    "@types/jest": "^25.1.4",
    "@types/node": "^13.9.3",
    "@types/ramda": "^0.26.18",
    "@types/react": "^16.9.25",
    "@types/react-intl": "^3.0.0",
    "@vtex/styleguide": "^9.78.10",
    "@vtex/test-tools": "^3.0.3",
    "@vtex/tsconfig": "^0.4.4",
    "apollo-cache-inmemory": "^1.6.5",
    "apollo-client": "^2.6.8",
    "typescript": "3.8.3"
  },
  "version": "2.8.3"
}
