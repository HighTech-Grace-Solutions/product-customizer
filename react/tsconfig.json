{
  "extends": "@vtex/tsconfig",
  "compilerOptions": {
    "lib": ["dom"],
    "resolveJsonModule": true,
    "types": ["node", "jest"],
    "module": "esnext",
    "moduleResolution": "node",
    "target": "es2017"
  },
  "include": ["./typings/*.d.ts", "./**/*.tsx", "./**/*.ts"]
}
