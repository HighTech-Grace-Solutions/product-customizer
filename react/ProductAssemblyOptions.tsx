import React, { FC, Fragment } from 'react'

import useAssemblyOptions from './modules/useAssemblyOptions'
import ProductAssemblyOptionsGroup from './components/ProductAssemblyOptions/ProductAssemblyOptionsGroup'
import { ProductAssemblyGroupContextProvider } from './components/ProductAssemblyContext/Group'

const ProductAssemblyOptions: FC = ({ children }) => {
  const assemblyOptions = useAssemblyOptions()

  if (!assemblyOptions) {
    return null
  }

  return (
    <Fragment>
      {Object.keys(assemblyOptions).map((assemblyOptionId) => (
        <ProductAssemblyGroupContextProvider
          key={assemblyOptionId}
          assemblyOption={assemblyOptions[assemblyOptionId]}
          name={assemblyOptions[assemblyOptionId].id}
        >
          <ProductAssemblyOptionsGroup>{children}</ProductAssemblyOptionsGroup>
        </ProductAssemblyGroupContextProvider>
      ))}
    </Fragment>
  )
}
export default ProductAssemblyOptions
