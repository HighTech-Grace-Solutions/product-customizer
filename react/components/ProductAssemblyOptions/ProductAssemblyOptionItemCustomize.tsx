import React, { Fragment, useState, FC } from 'react'
import { Button, Modal } from 'vtex.styleguide'
import ProductPrice from 'vtex.store-components/ProductPrice'
import { useDevice } from 'vtex.device-detector'
import { FormattedMessage } from 'react-intl'

import { useProductAssemblyItem } from '../ProductAssemblyContext/Item'
import ProductAssemblyOptionsGroup from './ProductAssemblyOptionsGroup'
import { imageUrlForSize } from './ProductAssemblyOptionItemImage'
import { ProductAssemblyGroupContextProvider } from '../ProductAssemblyContext/Group'
import { withItem } from './withItem'
import StopPropagation from '../StopPropagation'

const IMG_SIZE = 140

const ModalView: FC<{ closeAction: () => void }> = ({
  children,
  closeAction,
}) => {
  const {
    image,
    name,
    children: itemChildren,
    price,
  } = useProductAssemblyItem() as AssemblyItem

  return (
    <div className="flex flex-column">
      <div className="flex">
        {/* eslint-disable-next-line jsx-a11y/alt-text */}
        <img
          src={imageUrlForSize(image, IMG_SIZE)}
          width={IMG_SIZE}
          height={IMG_SIZE}
        />
        <div className="flex flex-column ml5">
          <span className="t-heading-5 c-on-base">{name}</span>
          <ProductPrice
            showLabels={false}
            showListPrice={false}
            sellingPriceContainerClass="t-heading-3 c-on-base"
            sellingPrice={price / 100}
          />
        </div>
      </div>
      {itemChildren &&
        Object.keys(itemChildren).map((childId) => {
          return (
            <ProductAssemblyGroupContextProvider
              key={childId}
              assemblyOption={itemChildren[childId]}
            >
              <ProductAssemblyOptionsGroup>
                {children}
              </ProductAssemblyOptionsGroup>
            </ProductAssemblyGroupContextProvider>
          )
        })}
      <div className="mt3">
        <Button block onClick={closeAction}>
          <div className="c-on-action-primary">DONE</div>
        </Button>
      </div>
    </div>
  )
}

interface Props {
  buttonProps?: ButtonProps
}

interface ButtonProps {
  collapse?: 'left' | 'right' | 'none'
}

const ProductAssemblyOptionItemCustomize: FC<Props> = ({
  children,
  buttonProps = {},
}) => {
  const {
    name,
    children: itemChildren,
  } = useProductAssemblyItem() as AssemblyItem

  const { isMobile } = useDevice()
  const buttonCollapse = buttonProps.collapse
  const [modalOpen, setModalOpen] = useState(false)

  if (!itemChildren) {
    return null
  }

  const handleClick = () => {
    setModalOpen(true)
  }

  const closeAction = () => setModalOpen(false)

  return (
    <Fragment>
      <Button
        variation="tertiary"
        onClick={handleClick}
        collapseLeft={buttonCollapse === 'left'}
        collapseRight={buttonCollapse === 'right'}
      >
        <div className="c-action-primary t-action">
          <FormattedMessage id="store/product-customizer.customize" />
        </div>
      </Button>
      <StopPropagation>
        <Modal
          isOpen={modalOpen}
          onClose={closeAction}
          centered={!isMobile}
          title={`Customize your ${name}`}
        >
          <ModalView closeAction={closeAction}>{children}</ModalView>
        </Modal>
      </StopPropagation>
    </Fragment>
  )
}

export default withItem(ProductAssemblyOptionItemCustomize)
