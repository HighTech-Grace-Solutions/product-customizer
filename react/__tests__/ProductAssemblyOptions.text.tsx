import React from 'react'
import { fireEvent, render, waitFor } from '@vtex/test-tools/react'
import useProduct, { ProductContext } from 'vtex.product-context/useProduct'

import ProductAssemblyOptions from '../ProductAssemblyOptions'
import InputValue from '../ProductAssemblyOptionInputValues'
import productSubscription from '../__fixtures__/productSubscription.json'

const mockUseProduct = useProduct as jest.Mock<ProductContext>

mockUseProduct.mockImplementation(() => ({
  product: productSubscription.data.product,
  selectedItem: productSubscription.data.product.items[0],
  selectedQuantity: 1,
}))

test('localizes subscription labels', async () => {
  const { getByText, container } = render(
    <ProductAssemblyOptions>
      <InputValue optionsDisplay="box" />
    </ProductAssemblyOptions>
  )

  const btn = getByText(/add subscription/i)

  fireEvent.click(btn)

  await waitFor(() =>
    expect(
      getByText(
        productSubscription.data.product.itemMetadata.items[0]
          .assemblyOptions[0].inputValues[0].defaultValue
      )
    ).toBeInTheDocument()
  )

  expect(container).toMatchSnapshot()
})
