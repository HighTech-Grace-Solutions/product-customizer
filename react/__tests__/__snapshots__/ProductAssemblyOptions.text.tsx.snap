// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`localizes subscription labels 1`] = `
<div>
  <div
    class="flex justify-between"
  >
    <div
      class="ttc-s pv4 c-muted-2 t-small"
    >
      Subscription
    </div>
    <div>
      <button
        class="vtex-button bw1 ba fw5 v-mid relative pa0 lh-solid br2 min-h-small t-action--small bg-transparent b--transparent c-action-primary hover-b--transparent hover-bg-action-secondary hover-b--action-secondary pointer "
        tabindex="0"
        type="button"
      >
        <div
          class="vtex-button__label flex items-center justify-center h-100 ph2 nr1 "
          style="padding-top: .25em; padding-bottom: .32em;"
        >
          Remove
        </div>
      </button>
    </div>
  </div>
  <div
    class="undefined mb4"
  >
    <div
      class="undefined mb3"
    >
      <span
        class="undefined c-muted-1 t-small overflow-hidden"
      />
    </div>
    <div
      class="undefined inline-flex flex-wrap flex items-center"
    >
      <div
        class="relative di pointer flex items-center outline-0 mr4"
        role="button"
        tabindex="0"
      >
        <div
          class="frameAround absolute b--action-primary br3 bw1 ba"
        />
        <div
          class="w-100 h-100 ba br2 bw1 b--muted-4 z-1 c-muted-5 flex items-center overflow-hidden"
        >
          <div
            class="c-on-base center pv3 ph5 z-1 t-body"
          >
            7 day
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="undefined mb4"
  >
    <div
      class="undefined mb3"
    >
      <span
        class="undefined c-muted-1 t-small overflow-hidden"
      />
    </div>
    <div
      class="undefined inline-flex flex-wrap flex items-center"
    >
      <div
        class="relative di pointer flex items-center outline-0 mr4"
        role="button"
        tabindex="0"
      >
        <div
          class="frameAround absolute b--action-primary br3 bw1 ba"
        />
        <div
          class="w-100 h-100 ba br2 bw1 b--muted-4 z-1 c-muted-5 flex items-center overflow-hidden"
        >
          <div
            class="c-on-base center pv3 ph5 z-1 t-body"
          >
            0
          </div>
        </div>
      </div>
      <div
        class="relative di pointer flex items-center outline-0 mr4"
        role="button"
        tabindex="0"
      >
        <div
          class="frameAround absolute b--action-primary br3 bw1"
        />
        <div
          class="w-100 h-100 ba br2 bw1 b--muted-4 z-1 c-muted-5 flex items-center overflow-hidden hover-b--muted-2"
        >
          <div
            class="c-on-base center pv3 ph5 z-1 t-body"
          >
             2
          </div>
        </div>
      </div>
      <div
        class="relative di pointer flex items-center outline-0 mr4"
        role="button"
        tabindex="0"
      >
        <div
          class="frameAround absolute b--action-primary br3 bw1"
        />
        <div
          class="w-100 h-100 ba br2 bw1 b--muted-4 z-1 c-muted-5 flex items-center overflow-hidden hover-b--muted-2"
        >
          <div
            class="c-on-base center pv3 ph5 z-1 t-body"
          >
             6
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
